src/java/org/apache/cassandra/db/virtual/CQLMetricsTable.java

```diff
@@ -0,0 +1,87 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.virtual;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import org.apache.cassandra.db.marshal.DoubleType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.dht.LocalPartitioner;
+import org.apache.cassandra.metrics.CQLMetrics;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.cql3.QueryProcessor;
+
+// Virtual table for CQL metrics
+public class CQLMetricsTable extends AbstractVirtualTable
+{
+    public static final String TABLE_NAME = "cql_metrics";
+    public static final String PREPARED_STATEMENTS_COUNT_COL = "prepared_statements_count";
+    public static final String PREPARED_STATEMENTS_EVICTED_COL = "prepared_statements_evicted";
+    public static final String PREPARED_STATEMENTS_EXECUTED_COL = "prepared_statements_executed";
+    public static final String PREPARED_STATEMENTS_RATIO_COL = "prepared_statements_ratio";
+    public static final String REGULAR_STATEMENTS_EXECUTED_COL = "regular_statements_executed";
+    public static final String NAME_COL = "name";
+    public static final String VALUE_ROW = "value";
+
+    private final CQLMetrics cqlMetrics;
+
+    // Default constructor references query processor metrics
+    CQLMetricsTable(String keyspace)
+    {
+        this(keyspace, QueryProcessor.metrics);
+    }
+
+    // For dependency injection
+    @VisibleForTesting
+    CQLMetricsTable(String keyspace, CQLMetrics cqlMetrics)
+    {
+        // create virtual table with this name
+        super(TableMetadata.builder(keyspace, TABLE_NAME)
+                           .kind(TableMetadata.Kind.VIRTUAL)
+                           .partitioner(new LocalPartitioner(UTF8Type.instance))
+                           .addPartitionKeyColumn(NAME_COL, UTF8Type.instance)
+                           .addRegularColumn(PREPARED_STATEMENTS_COUNT_COL, DoubleType.instance)
+                           .addRegularColumn(PREPARED_STATEMENTS_EVICTED_COL, DoubleType.instance)
+                           .addRegularColumn(PREPARED_STATEMENTS_EXECUTED_COL, DoubleType.instance)
+                           .addRegularColumn(PREPARED_STATEMENTS_RATIO_COL, DoubleType.instance)
+                           .addRegularColumn(REGULAR_STATEMENTS_EXECUTED_COL, DoubleType.instance)
+                           .build());
+        this.cqlMetrics = cqlMetrics;
+    }
+
+    @Override
+    public DataSet data()
+    {
+        // populate metrics in the virtual table
+        SimpleDataSet result = new SimpleDataSet(metadata());
+        addRow(result, VALUE_ROW, cqlMetrics);
+
+        return result;
```

Comment:
minor nit: it might be worth considering inlining since we don't call `addRow` anywhere else?
```suggestion
        return new SimpleDataSet(metadata()).row(VALUE_ROW)
                                            .column(PREPARED_STATEMENTS_COUNT_COL, cqlMetrics.preparedStatementsCount.getValue())
                                            .column(PREPARED_STATEMENTS_EVICTED_COL, cqlMetrics.preparedStatementsEvicted.getCount())
                                            .column(PREPARED_STATEMENTS_EXECUTED_COL, cqlMetrics.preparedStatementsExecuted.getCount())
                                            .column(PREPARED_STATEMENTS_RATIO_COL, cqlMetrics.preparedStatementsRatio.getValue())
                                            .column(REGULAR_STATEMENTS_EXECUTED_COL, cqlMetrics.regularStatementsExecuted.getCount());
```

Reply:
This is the pattern used in all virtual table classes, so like to keep inline with them.

Reply:
With inverted table code, this comment no longer applicable